packages/framer-motion/src/render/VisualElement.ts

```diff
@@ -0,0 +1,810 @@
+import sync, { cancelSync } from "framesync"
+import { invariant } from "hey-listen"
+import { createElement } from "react"
+import { MotionConfigProps } from "../components/MotionConfig"
+import { ReducedMotionConfig } from "../context/MotionConfigContext"
+import { SwitchLayoutGroupContext } from "../context/SwitchLayoutGroupContext"
+import { featureDefinitions } from "../motion/features/definitions"
+import { FeatureBundle, FeatureDefinition } from "../motion/features/types"
+import { MotionProps, MotionStyle } from "../motion/types"
+import { Box } from "../projection/geometry/types"
+import { IProjectionNode } from "../projection/node/types"
+import { TargetAndTransition } from "../types"
+import { isRefObject } from "../utils/is-ref-object"
+import { env } from "../utils/process"
+import { initPrefersReducedMotion } from "../utils/reduced-motion"
+import {
+    hasReducedMotionListener,
+    prefersReducedMotion,
+} from "../utils/reduced-motion/state"
+import { SubscriptionManager } from "../utils/subscription-manager"
+import { motionValue, MotionValue } from "../value"
+import { isWillChangeMotionValue } from "../value/use-will-change/is"
+import { isMotionValue } from "../value/utils/is-motion-value"
+import {
+    ResolvedValues,
+    VariantStateContext,
+    VisualElementEventCallbacks,
+    VisualElementOptions,
+} from "./types"
+import { AnimationState, variantPriorityOrder } from "./utils/animation-state"
+import {
+    isControllingVariants as checkIsControllingVariants,
+    isVariantNode as checkIsVariantNode,
+} from "./utils/is-controlling-variants"
+import { isVariantLabel } from "./utils/is-variant-label"
+import { updateMotionValuesFromProps } from "./utils/motion-values"
+import { resolveVariantFromProps } from "./utils/resolve-variants"
+
+const featureNames = Object.keys(featureDefinitions)
+const numFeatures = featureNames.length
+
+const propEventHandlers = [
+    "AnimationStart",
+    "AnimationComplete",
+    "Update",
+    "Unmount",
+    "BeforeLayoutMeasure",
+    "LayoutMeasure",
+    "LayoutAnimationStart",
+    "LayoutAnimationComplete",
+]
```

Comment:
Ah, we are. I think we can do
```suggestion
const propEventHandlers = [
    "AnimationStart",
    "AnimationComplete",
    "Update",
    "Unmount",
    "BeforeLayoutMeasure",
    "LayoutMeasure",
    "LayoutAnimationStart",
    "LayoutAnimationComplete",
] as const
```
here to improve the type

Reply:
Or, event better:
```suggestion
const propEventHandlers: (keyof VisualElementEventCallbacks)[] = [
    "AnimationStart",
    "AnimationComplete",
    "Update",
    "Unmount",
    "BeforeLayoutMeasure",
    "LayoutMeasure",
    "LayoutAnimationStart",
    "LayoutAnimationComplete",
]
```

Reply:
I've just added this, is this saying the array doesnt change?

Reply:
`as const` makes sure it is not typed as `string[]`, but a tuple of these string values